### ISSUE TITLE
Modifying the AST from python will create node copies that are lost

### ISSUE THREAD

**Issue by ohm314** [BlueBrain/nmodl#702]
See the new test case here:

https://github.com/BlueBrain/nmodl/blob/c159821b2a73ad23416d1364727dd1ce279b46ba/test/unit/pybind/test_visitor.py#L94-L106

When we use a visitor to modify the AST (e.g. replace a node) from python the changes are lost. The AST nodes in the `vist_` functions seem to be copies of the "original" objects and changes on them are lost as we recurse out of the tree.

This seems to be a regression that was introduced when the visitors were changed to take node references rather than pointers. I confirmed that the issue is not present at version 0.2.

**Comment by cattabiani** [comment]
once it is solved we should add a unit test to avoid repeating this error in the future

**Comment by ohm314** [comment]
That unit test is what I refer to above! Thought would be good to start with that :)

Here's my work in progress: https://github.com/BlueBrain/nmodl/tree/fix/pyvst_lost_nodes
I've added a PtrVisitor class that takes pointers instead of references. I don't know, there might be a better way - one should maybe setup a tiny example with pybind11 and two classes and try to reproduce it there and find a better fix.

At any rate, my current code is not working yet. When you try to import `nmodl` you get:

```
generic_type: type "AstLookupVisitor" referenced unknown base type "nmodl::visitor::Visitor"
```

So there is a bit of a mess with the various visitors now and the way they inherit from each other.

**Comment by alkino** [comment]
Fixed by #708
